support/man/strlcpy.3
.\" $OpenBSD: strlcpy.3,v 1.15 2003/04/12 21:56:39 millert Exp $
.\"
.\" Copyright (c) 1998, 2000 Todd C. Miller <Todd.Miller@courtesan.com>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND TODD C. MILLER DISCLAIMS ALL
.\" WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL TODD C. MILLER BE LIABLE
.\" FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
.\" OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
.\" CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd June 22, 1998
.Dt STRLCPY 3
.Os
.Sh NAME
.Nm strlcpy ,
.Nm strlcat
.Nd size-bounded string copying and concatenation
.Sh SYNOPSIS
.Fd #include <string.h>
.Ft size_t
.Fn strlcpy "char *dst" "const char *src" "size_t size"
.Ft size_t
.Fn strlcat "char *dst" "const char *src" "size_t size"
.Sh DESCRIPTION
The
.Fn strlcpy
and
.Fn strlcat
functions copy and concatenate strings respectively.
They are designed
to be safer, more consistent, and less error prone replacements for
.Xr strncpy 3
and
.Xr strncat 3 .
Unlike those functions,
.Fn strlcpy
and
.Fn strlcat
take the full size of the buffer (not just the length) and guarantee to
NUL-terminate the result (as long as
.Fa size
is larger than 0 or, in the case of
.Fn strlcat ,
as long as there is at least one byte free in
.Fa dst ) .
Note that you should include a byte for the NUL in
.Fa size .
Also note that
.Fn strlcpy
and
.Fn strlcat
only operate on true
.Dq C
strings.
This means that for
.Fn strlcpy
.Fa src
must be NUL-terminated and for
.Fn strlcat
both
.Fa src
and
.Fa dst
must be NUL-terminated.
.Pp
The
.Fn strlcpy
function copies up to
.Fa size
- 1 characters from the NUL-terminated string
.Fa src
to
.Fa dst ,
NUL-terminating the result.
.Pp
The
.Fn strlcat
function appends the NUL-terminated string
.Fa src
to the end of
.Fa dst .
It will append at most
.Fa size
- strlen(dst) - 1 bytes, NUL-terminating the result.
.Sh RETURN VALUES
The
.Fn strlcpy
and
.Fn strlcat
functions return the total length of the string they tried to create.
For
.Fn strlcpy
that means the length of
.Fa src .
For
.Fn strlcat
that means the initial length of
.Fa dst
plus
the length of
.Fa src .
While this may seem somewhat confusing it was done to make
truncation detection simple.
.Pp
Note however, that if
.Fn strlcat
traverses
.Fa size
characters without finding a NUL, the length of the string is considered
to be
.Fa size
and the destination string will not be NUL-terminated (since there was
no space for the NUL).
This keeps
.Fn strlcat
from running off the end of a string.
In practice this should not happen (as it means that either
.Fa size
is incorrect or that
.Fa dst
is not a proper
.Dq C
string).
The check exists to prevent potential security problems in incorrect code.
.Sh EXAMPLES
The following code fragment illustrates the simple case:
.Bd -literal -offset indent
char *s, *p, buf[BUFSIZ];

\&...

(void)strlcpy(buf, s, sizeof(buf));
(void)strlcat(buf, p, sizeof(buf));
.Ed
.Pp
To detect truncation, perhaps while building a pathname, something
like the following might be used:
.Bd -literal -offset indent
char *dir, *file, pname[MAXPATHLEN];

\&...

if (strlcpy(pname, dir, sizeof(pname)) >= sizeof(pname))
	goto toolong;
if (strlcat(pname, file, sizeof(pname)) >= sizeof(pname))
	goto toolong;
.Ed
.Pp
Since we know how many characters we copied the first time, we can
speed things up a bit by using a copy instead of an append:
.Bd -literal -offset indent
char *dir, *file, pname[MAXPATHLEN];
size_t n;

\&...

n = strlcpy(pname, dir, sizeof(pname));
if (n >= sizeof(pname))
	goto toolong;
if (strlcpy(pname + n, file, sizeof(pname) - n) >= sizeof(pname) - n)
	goto toolong;
.Ed
.Pp
However, one may question the validity of such optimizations, as they
defeat the whole purpose of
.Fn strlcpy
and
.Fn strlcat .
As a matter of fact, the first version of this manual page got it wrong.
.Sh SEE ALSO
.Xr snprintf 3 ,
.Xr strncat 3 ,
.Xr strncpy 3



CHANGES
 
  Copyright (c) 1999-2003 WU-FTPD Development Group.  
  All rights reserved.
  
  Portions Copyright (c) 1980, 1985, 1988, 1989, 1990, 1991, 1993, 1994
    The Regents of the University of California.
  Portions Copyright (c) 1993, 1994 Washington University in Saint Louis.
  Portions Copyright (c) 1996, 1998 Berkeley Software Design, Inc.
  Portions Copyright (c) 1989 Massachusetts Institute of Technology.
  Portions Copyright (c) 1998 Sendmail, Inc.
  Portions Copyright (c) 1983, 1995, 1996, 1997 Eric P.  Allman.
  Portions Copyright (c) 1997 Stan Barber.
  Portions Copyright (c) 1997 Kent Landfield.
  Portions Copyright (c) 1991, 1992, 1993, 1994, 1995, 1996, 1997
    Free Software Foundation, Inc.  
 
  Use and distribution of this software and its source code are governed 
  by the terms and conditions of the WU-FTPD Software License ("LICENSE").
 
  If you did not receive a copy of the license, it may be obtained online
  at http://www.wu-ftpd.info/license.html.
 
  $Id: CHANGES,v 1.47 2016/03/14 23:57:54 wmaton Exp $

###########################################################################

Changes in 2.8.0-CC5: Released

 o  Updates to config.guess and config.sub to modernise the autoconf
    build system.

 o  configure script regenerated.

 o  Describe auto-buf tuning configure option in README.AUTOCONF.

 o  Add option (-R) to disable REST command in cases where download
    managers can abuse the server by opening multiple connections to
    retrieve chunks of the same file.  Based on a patch by Philippe Troin
    (See Debian bug ID 138497).

 o  Some minor man page cleanups.

 o  More syncs to Debian patches dating back to 2.4.2.

 o  Update INSTALL instructions.

 o  Changed the TLS default cipher from ALL:!EXP to exclude most insecure
    or unsupported ciphers within modern versions of OpenSSL.

Changes in 2.8.0-CC4: Released 7 January, 2012

 o  Thanks to a patch by Laurent CONSTANTIN, calling SSL_library_init()
    will no longer create an error within the SSL code related to a lack
    of ciphers.  The patch is applied to src/tlsutil.c.

 o  Another patch by Laurent CONSTANTIN, this time eliminating extra
    carriage returns that aren't necessary when calling syslog().

 o  Added -c option which displays the compile-time options that were used
    to build WU-FTPD. 

Changes in 2.8.0-CC3: Released 6 November, 2011

 o  Many URL updates.

 o  The default email address used by the server in absence of an email
    directive in ftpaccess is ftp-bugs, and is noted in ftpaccess now.

 o  Added code for ADLER-32 checksums for future extension to the list
    of algorithms available to do checksums.

 o  hard-coded links to cksum and md5sum within the FTP root now point to
    these utilities in /bin instead of /usr/bin.

 o  As defined in GFD.20 (GridFTP:  Protocol Extension to FTP for the
    Grid) by the Global Grid Forum, the SBUF command has been introduced.
    This calls the same code as SITE BUFSIZE would.  SITE BUFSIZE is
    still present for backwards compatibility.

 o  Thanks to a contribution by Austin Ellis, added RHEL6 packaging info
    to the codebase in packages/rpm.  This also includes a patch for
    building on that particular system.

 o  Updated ftpd.8 manpage documenting some missing SITE commands that
    had been in the server since 1998.

 o  First cut at implementing the remainder of the RFC3659 extended FTP
    commands.  FEAT, MLST, OPTS and MLSD are now supported.  Based upon
    early gridFTP releases which used WU-FTPD 2.6.0 and 2.6.1 code.

 o  According to draft-somers-ftp-mfxx-04.txt, defined the stubs for the
    MFF, MFCT and MFMT commands.  These may or may not be supported in the
    future, but at the least they are defined for now.  If so, they may
    become SITE commands instead if the spec is not approved.

 o  Updated doc/examples/ftpconversions based on Debian changes introduced
    sometime in 1995.  The server can now deal with such things as bzip2,
    xz, zip and unzip.

 o  If there's an empty /etc/fstab, a segfault can be produced.  A fix
    from Debian BugID 568235 to path_to_device in src/extensions.c addresses
    this.  Patch by Chris Butler.

 o  Added another possible path for quota.h, patch by Chris Butler of
    Debian.

 o  Added missing handler for SIGTERM, patch by Chris Butler of Debian
    related to Debian bug #536774.

 o  Changed -h invocation on the commandline to print out a very brief help
    listing all commandline options that could be supplied to the daemon.
    Also note that some switches are or have an effect on the configuration
    file.  The former function assigned to the 'h' option is now 'H'.

 o  The former function assigned to the 'h' option is now 'H'.  Said it
    twice now.  The manpage has been updated to reflect the changes.

 o  Implemented the CSID command as found in the IETF draft
    draft-peterson-streamlined-ftp-command-extensions.  This is meant
    to be a standards-based replacement for the CLNT command.  Currently
    this command is hard-wired with the server name and version, but the 
    next release will permit customisation.

###########################################################################

Changes in 2.8.0-CC2: Released 26 June, 2009

 o  Cleaned-up and updated the ftpd.8 manpage which had languished for
    some time.

 o  Added Auto tuning/auto buf, with added commands SITE BUFSIZE and
    SITE BUFSIZEMEASURE.  Use --enable-autobuf to enable this feature.
    Based on a patch on work done on WU-FTPD-2.6.1 by Gaurav Navlakha, et al
    of DAST, NLANR (now CAIDA).  This code also contains the ability to
    report back on buffer usage.  Note that an autobuf client needs to
    be used to take advantage of this feature.  For the time being, the
    code for buffer measurement is IPv4 only.

 o  Updated src/extensions.c to take into account a change in Linux quota
    structure.  Based on a patch by Chris Butler.


support/strlcpy.c
/*	$OpenBSD: strlcpy.c,v 1.7 2003/04/12 21:56:39 millert Exp $	*/

/*
 * Copyright (c) 1998 Todd C. Miller <Todd.Miller@courtesan.com>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND TODD C. MILLER DISCLAIMS ALL
 * WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL TODD C. MILLER BE LIABLE
 * FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
 * OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#if defined(LIBC_SCCS) && !defined(lint)
static char *rcsid = "$OpenBSD: strlcpy.c,v 1.7 2003/04/12 21:56:39 millert Exp $";
#endif /* LIBC_SCCS and not lint */

#include <sys/types.h>
#include <string.h>

/*
 * Copy src to string dst of size siz.  At most siz-1 characters
 * will be copied.  Always NUL terminates (unless siz == 0).
 * Returns strlen(src); if retval >= siz, truncation occurred.
 */
size_t
strlcpy(char *dst, const char *src, size_t siz)
{
	register char *d = dst;
	register const char *s = src;
	register size_t n = siz;

	/* Copy as many bytes as will fit */
	if (n != 0 && --n != 0) {
		do {
			if ((*d++ = *s++) == 0)
				break;
		} while (--n != 0);
	}

	/* Not enough room in dst, add NUL and traverse rest of src */
	if (n == 0) {
		if (siz != 0)
			*d = '\0';		/* NUL-terminate dst */
		while (*s++)
			;
	}

	return(s - src - 1);	/* count does not include NUL */
}

